teps/0032-tekton-notifications.md

```diff
@@ -0,0 +1,167 @@
+
+## Motivation
+
+Tekton is a tool for building CI/CD systems. It allows automating workflows
+using pipelines which may run autonomously most of the times, until something
+goes wrong or they reach as step that require human involvement.
+
+Notification can help raising the attention of an human being to a specific
+workflow execution that is need of attention.
+
+Several Tekton users ask about notifications in Tekton. We have developed two
+features that may help implementing notifications:
+
+- [Cloud events](https://tekton.dev/docs/pipelines/events/#events-via-cloudevents)
+- [Finally](https://tekton.dev/docs/pipelines/pipelines/#adding-finally-to-the-pipeline)
+
+Cloud events are triggered every time a pipeline or task is started, begins
+executing and is terminated. Cloud events, combined with Tekton Triggers can
+be used to execute tasks and pipelines as a reaction to an event. This is
+used in our dogfooding cluster to generate notifications for failed CD pipelines.
+It works fine, however it requires a good amount of YAML and plumbing resource
+together to achieve the goal of notifications.
+
+The "Finally" section in a pipeline allows execution tasks once all regular tasks
+completed execution. Such "final" tasks can be used to send notifications about
+the outcome of a pipeline, as they will always be invoked, even if a task in the
+main pipeline failed. However final tasks are only executed at the end, so they
+cannot be used for "start" notifications, and they are limited to pipelines.
+Work is in progress to allow "final" tasks access to the execution status of tasks
+in the main pipeline.
+
+Tekton users would highly benefit from native notification support.
+
+### Goals
+
+The main goal of the TEP is to design and implement a notification system that
+is both easy to use and flexible, for delivering notification targeted to humans.
+
+Notifications are easy to use: enabling and configuring default notifications
+to get something up and running (notifying on every event of every Pipeline/Task)
+requires only few lines of configuration.
+
+Notifications are flexible: users are able to configure the precise Pipelines/Tasks
+and events that notifications are sent for. Users can also modify the configuration
+for their chosen notification system.
+
+A secondary goal is to make sure that notification can benefit from our growing
+catalog of tasks, several of which have been written to deliver notification through
+different platforms.
+
+### Non-Goals
+
+- Build a machine to machine communication protocols
+- Build pre/post execution hook for steps, tasks or pipelines
+- Build a catalog of integration with different notification, chat or email systems
+- Build a system to send large amounts of data produced by a task or pipeline
+- Write a controller for cloud events (even if that could be part of the solution).
+  See [tektoncd/pipeline#2944](https://github.com/tektoncd/pipeline/issues/2944) for
+  further context.
```

Comment:
Up to you: add a non-goal about creating an event bus system in Tekton (e.g. using Tekton notifications in a pipeline to explicitly trigger or change other things in the pipeline). Notifications should be events from Tekton to something else.